src/CBot/CBotClass.cpp

```diff
@@ -444,18 +459,29 @@ bool CBotClass::CheckCall(CBotProgram* program, CBotDefParam* pParam, CBotToken*
 ////////////////////////////////////////////////////////////////////////////////
 CBotClass* CBotClass::Compile1(CBotToken* &p, CBotCStack* pStack)
 {
-    if ( !IsOfType(p, ID_PUBLIC) )
+    CBotProgram* program = pStack->GetProgram();
+
+    bool isPrivate;
+
+    if ( IsOfType(p, ID_PUBLIC) ) isPrivate = false;
+    else if ( IsOfType(p, ID_PRIVATE) ) isPrivate = true;
+    else if ( IsOfType(p, ID_PROTECTED) )
     {
-        pStack->SetError(CBotErrNoPublic, p);
+        pStack->SetError( CBotErrBadPrivilege, p );
+        return nullptr;
+    }
+    else
+    {
+        pStack->SetError( CBotErrNoPrivilege, p );
         return nullptr;
     }
 
-    if ( !IsOfType(p, ID_CLASS) ) return nullptr;
+    IsOfType(p, ID_CLASS); // skip class keyword
 
-    std::string name = p->GetString();
+    std::string name = p->GetString(); // name of class
 
-    CBotClass* pOld = CBotClass::Find(name);
-    if ( pOld != nullptr && pOld->m_IsDef )
+    if ( (isPrivate && program->ClassExists(name)) ||            // check only in current program
+         (!isPrivate && CBotClass::Exists(name, program)) )      // check in current program and list of public classes
     {
```

Comment:
I thought it would be useful to be able to overwrite public class with the same name form other program by private class, but in fact it can be deceptive.